### DISCUSSION TITLE
Next steps in improving the documentation system.

### DISCUSSION THREAD

**Discussion by mmatera** [Mathics3/mathics-core#1023] [General]
Now that mathics-core works again without issues in the CI, I continued working on the documentation system.

Among the pending tasks

  - [x]  (a.) Fix some broken links in the documentation.
  - [x]  (b.) DRY code in `mathics.doc` and `mathics.docpipeline`.
  - [x]  (c.) Start to split the code in `mathics.doc.common_doc` in a part related to the organization of the code, and a part
         related to parsing the documentation entries and organizing the doctests.
  - [x]  (d.) More comments, documentation, and tests for the documentation system.
  - [ ]  (e.) avoiding go over to the full documentation system to access a specific set of tests.
  - [ ]  (f.) Allow pymathics modules to use the docpipeline system to tests their symbols.
  - [ ]  (g.) restore the proper ordering of chapters and sections.
  - [ ]  (h.) review the structure of the documentation.
  - [ ]  (i.) check that both mathics-django and latex documentation looks as it should look.

...  (add other items that you think must be present in this list )

  - [ ]  (y.) Split the documentation system as a separate package.
  - [ ]  (z.) From the homebrew documentation system, produce some more standard documentation like sphinx or at least markdown documentation.

Today I think made some progress in the first of these points

#1021  addresses a. to d.
#1022  addresses partially e.  and f.
#1025 goes further in c., g., h. and i.

In order to continue with this task, I think we need to discuss some details about the structure of the documentation. I will try to start the discussion here.

Guide Sections and builtin packages
-----------------------------------

Several of the "FIXME" comments inside the code of the documentation system are related to the way in which we are handiling packages (subfolders) inside ``mathics.builtin`` and pymathics modules. In the current code, each file and folder inside `mathics.builtin` and `pymathics` modules are mapped to "chapters" inside a part ("Reference of Built-in Symbols" and "Mathics3 Modules" respectively).

For files inside these modules, the module docstring is used as the introduction to the chapter, which is followed by a Chapter index and a set of sections corresponding to each builtin symbol defined in the module.

For packages inside the top-level modules, the docstring in the `__init__.py` file is used to generate the introductory text of the chapter, which contains
a "guide section" and "sections". The "guide section" starts with a documentation chunk containing exactly the same than the chapter introduction,
and "subsections", which maps the list of sub-modules. A the level of the chapter, after the "guide section" we find (regular) "sections", with are in correspondece with the submodules. The (regular) sections under a chapter have an initial text, built from the sub-module docstring, and "subsections" comming from the symbols defined inside the corresponding submodule.

To make this galimatias more clear, let's see two examples.

* From `mathics.builtin.procedure` (a single file module) we produce

```
   Chapter "Procedural":
      introduction text (from ``mathics.builtin.procedure.__doc__``)
      chapter index
      section "Abort" (from the docstring of the class  `mathics.builtin.procedure.Abort`)
      section "Break" (from the docstring of the class  `mathics.builtin.procedure.Break`)
      ...
```

* From `mathics.builtin.colors`
```
   Chapter "Colors":
       introduction text (from ``mathics.builtin.colors.__init__.__doc__``)
       chapter index
       Guide section `Color`
            introduction text (again from ``mathics.builtin.colors.__init__.__doc__``)
	    subsection "Color Directives" (from ``mathics.builtin.colors.color_directives.__doc__``) (type section)
	    subsection "Named colors" (from ``mathics.builtin.colors.named_colors.__doc__``)         (type section)

       Section "Color Directives" (from ``mathics.builtin.colors.color_directives.__doc__``)
            subsection "CMYKColor" (docstring from the class ``mathics.builtin.colors.color_directives.CMYKColor``)  (type subsection)
	    subsection "ColorDistance" (docstring from the class ``mathics.builtin.colors.color_directives.ColorDistance``) (type subsection)
	    ...

       Section "Named colors" (from ``mathics.builtin.colors.named_colors.__doc__``)
            subsection "Black" (docstring from the class ``mathics.builtin.colors.named_colors.Black``) (type subsection)
	    subsection "Blue" (docstring from the class ``mathics.builtin.colors.named_colors.Blue``)   (type subsection)
	    ...
```

Notice that with the current organization,

- subsections of a Guide section are of class "DocSection", while in Sections are "DocSubsections".
- In docpipeline "Section" is used as an equivalent of "Symbol".
- In the LaTeX documentation, "Sections" and "Subsections" are shown different. But then, the symbol ``Black`` and symbol ``Break`` are shown in different formats.
- subsections in a Guide Section actually are pointers to the other sections in the chapter. Then, at the time, a subsection of a guide section has subsections.
- In the docpipeline, since the same documentation entry appears in two places (as a subsection of a subsection in the guide, and as a subsection of the chapter, and the documentation entry is also repeated as the chapter introduction and the Guide Section introduction), we have to actively avoid to process the same test two times.
- If a submodule has a submodule, "__init__.__doc__"  is loaded as another section, but its subsections are not loaded.

These inconsistencies both in the structure and in the formatting of the documentation must be removed in order to make the code simpler, and the documentation more readable and compatible with other documentation systems.

**Comment by mmatera** [comment]
Guides in WR
---------------------

WR is a quite sophisticated documentation structure, which has different kinds of entries. Some of the main kinds of entries are:

* Functions (Symbol entries): contain the documentation about a symbol.
* Guides: A description of a certain set of related symbols, and a list of links to the corresponding symbol entries. The links at the time are organized in subgroups by specific tasks, having their own guides.
*  Technotes, Tutorials, etc, etc.

The format is hypermedia and not hierarchic: two guides can include links to the same symbols, and there is no organization in chapters/ sections/ subsections. This scheme is OK for an online documentation system, but probably not for a book. 

Proposal to translate this to the Mathics Documentation System
-----------------------------------------------------------------------------------------------

A way to translate this to our documentation system would be

* Each module inside builtin/pymathics, at any level -> a DocChapter.  
* If the module is a single file, it contains sections, one for each symbol inside.
* If the module is a package, the chapter includes a guide for each submodule, and a section for each symbol on each submodule. 
* Each Part can contain a list of the top-level Modules/chapters to generate the entry point in the online documentation.


With this pattern,  each DocEntry appears exactly once. Thoughts?

**Reply by mmatera** [reply]
The proposal is implemented in  #1025. 

@rocky  I am going to leave this for a while until I get some feedback.

**Reply by rocky** [reply]
I looked at #1025 and I think the intent is good in terms of making a better distinction between the data structures used for Guide Sections versus Sections. and organizing the accumulation of data so that we can avoid having to look for duplicates.

That part is good, although I can't say I fully understand the code. But I am fine with going with something like this and tweaking later, if something arises.

There is some stuff , however, that is a bit weird though in formatting the LaTeX doc. The unnumbered sections like for `Accumulate[]` and `CubeRoot[]` are weird. Is the intent to try to deal with the DAG (directed acyclic graph) or non-tree structure that occurs as a result of a function belonging in several categories?  

![mathics3-arith-fns](https://github.com/Mathics3/mathics-core/assets/8851/2c93cb74-5e54-4348-92c8-e8b7f9f10430)


Whether that is the situation or not, please let's not get creative here and just follow the way books are typically written (including the Mathematica books): they are strictly heraricachal. And sections and subsections are numbered.

Of course, the way code is written in terms of modules is also hierarchal too. And therefore, unless there is good reason not to, having the hierarchy of the reference manual and the hierarchy of helps people find things in the code and  is natural. 

The problem of how to show tech sections and guide sections that allow for shared builtin functions is a separate problem. I don't think it a particularly _difficult_  problem to solve either, but at any rate that is a feature.  Let's please put off trying to get all of that implemented or visible in Django and just stay with a strictly hierarchical format for now, that follows the hierarchical nature of the book (and code).

**Reply by mmatera** [reply]
@rocky 
most of the changes are just moving functions from `mathics.doc.common` to `mathics.doc.structure` and `mathics.doc.gather`. The main change is that some of the methods inside `mathics.doc.structure.Part` were converted into non-member functions.

How the documentation is built from the mathics modules is determined by   `doc_chapter`, `gather_sections` and `gather_guides_and_sections` functions. I guess that this is the part that requires more rework.

In all of this I tried to be as conservative as possible, trying to follow the existent pattern. Still, there are some decisions to make, and some problems to fix. 
Regarding the numeration of the sections, I kept the  previous behavior which does not put numeration. This show another problem: there is no way to add page number references in the existent documentation syntax.

**Reply by mmatera** [reply]
Regarding 
> Whether that is the situation or not, please let's not get creative here and just follow the way books are typically written (including the Mathematica books): they are strictly heraricachal. And sections and subsections are numbered.

Is not a matter of creativity, but just the way in which the documentation is currently organized. One problem I found is that "Symbols" are identified with "Sections" in docpipeline, but with the organization in master, they could be "Section" or "Subsection" depending on the level of the module in which are defined. In my opinion, this second behavior is wrong, both because the format of the same kind of thing would be different, and because is not scalable for nested packages.

**Reply by rocky** [reply]
What does any of this have to do with the LaTeX documentation?

**Reply by mmatera** [reply]
LaTeX documentation is built from this abstract representation. The abstract structure partially defines how the LaTeX documentation looks

**Reply by rocky** [reply]
> The abstract structure partially defines how the LaTeX documentation looks

As I wrote previously, I don't  object to the underlying structure and code changes. I do object to the part that changes the way LaTeX documents come out. And for now we should keep Django online docs the same as well. Later, we can discuss how to do fancier guide sections technical guides and so on. For now, a simple hierarchical display is fine for Django.

**Reply by mmatera** [reply]
The only change here is in the content of the guide sections, which now, instead of producing copies of the symbol sections as subsections, has an index and a documentation block from the docstring of the corresponding package.

This does not have anything to do with the particular format of the latex documentation or the Django documentation. The structural elements are the same as before, and the difference is just on how the elements are produced from the python packages.

Changes in the formatting of different documentation elements are minimal and can be easily reverted, or reformulated.

**Reply by rocky** [reply]
Great! Please revert the LaTeX formatting changes. This is the part that is "creative" and not like any book.

**Reply by mmatera** [reply]
@rocky, the only changes in the formatting of LaTeX content consist of including the guides when a chapter is rendered and avoiding including a TOC inside a section guide. The rest of your comments about the format also apply to the documentation generated before all recent changes.

For example, this is how the Arithmetic first page looks once the changes in `latex_doc` are reverted:

![imagen](https://github.com/Mathics3/mathics-core/assets/2862288/4bd68746-e849-42ef-be77-27a80740be62)

The main change is that guides are not shown.

**Reply by rocky** [reply]
This is better. Thanks. 

Putting in Chapter and Section number would probably better and more like the way normal books such as the Mathematica books.

**Reply by mmatera** [reply]
On the other hand, this is how it looks by asking that sections have section numbers:
![imagen](https://github.com/Mathics3/mathics-core/assets/2862288/f917a337-f76b-4eff-90a8-d98e446cd3f4)

**Reply by rocky** [reply]
I like something in between where the flow is like a book. So instead of the list under Basic Arithmetic, the actual subsections appear. 

In Django or a online hyperlinked format, the unnumbered and listed section is fine which is the way Django documentation currently works. 

Thanks.

**Reply by mmatera** [reply]
Independently on the details, the main "new" idea in #1025 is to change the organization
of the documentation to this:
![imagen](https://github.com/Mathics3/mathics-core/assets/2862288/85e16038-af5f-41e2-ac4c-1c3e96bd25c0)


with documentation of  "Symbols" always translated to "Sections" in a chapter, while "Guide Sections" are just indices like in WR.

**Reply by mmatera** [reply]
With [8001c8b](https://github.com/Mathics3/mathics-core/pull/1025/commits/8001c8b4f2efdb953763e165192e53c7f866266f),
the chapters look like this:
![imagen](https://github.com/Mathics3/mathics-core/assets/2862288/87f8aae9-9cc1-425f-beef-46f2506bb481)

Django (with [#203](https://github.com/Mathics3/mathics-django/pull/203#issue-2194926238)) produces

In the **Arithmetic Functions** chapter
![Chapter index](https://github.com/Mathics3/mathics-core/assets/2862288/44aa4781-16fd-44ba-ad2e-e8f63f8509a7)

In the **Basic Arithmetic** guide
![Guide](https://github.com/Mathics3/mathics-core/assets/2862288/d9f35a04-5087-44e8-9841-81bd3f3e983b)

And in the **CubeRoot** section
![Section](https://github.com/Mathics3/mathics-core/assets/2862288/d931811a-ed55-4a03-8fde-00f0e9235ab7)


Again, both LaTeX and Django require some formatting fixes, but they must be implemented according to the basic structure of the documentation.

**Reply by rocky** [reply]
Guide sections should have sections in them. Possibly in the future all sections may be under some guide section, but this is not the situation now.  For the LaTeX doc under "Arithmetic Functions", the organization should be:

1. Arithmetic Functions
  1.1 Basic Arithmetic
     1.1.1 CubeRoot
     1.1.2 Divide
     1.2.3 Plus

The 5th edition of Mathematica also uses a 3-part sectioning system as do most books. 

The format of the "Contents" part of https://github.com/Mathics3/mathics-core/discussions/1023#discussioncomment-8951355 as a table with clickable page number is great, provided the sections get the numbing as shown above. 

The part with lists is a new phenomenon that should not exist in the LaTeX doc. In the Django online doc, that is fine and shows the summary text and these are links that are clickable.

**Reply by mmatera** [reply]
OK, then we arrive at what I tried to discuss in https://github.com/Mathics3/mathics-core/discussions/1023#discussioncomment-8801913

This discussion affects mainly the module  `mathics.doc.gather` in #1025, while `mathics.doc.structure` (the hierarchy of documentation parts/elements) is more related to how these elements link among them and are translated to different forms (HTML documentation, books, sets of tests, etc). Apart from cosmetic details (if we want indices like tables, how are the content distributed along the page) I guess that we achieve certain agreement on what these elements (Documentation, Part, Chapter, Section, and Subsections) are. 

So the following is about how to translate `mathics.builtin` / `pymathics` into this structure.


How should modules fit in the documentation hierarchy?
===============================================

(how `mathics.doc.gather` must be implemented)

In the old days pre mathics-code, `mathics.builtin` was a "shallow" package: it does not have subfolders. Each module has a set of symbols, and then the map "module->chapter", "symbol->section" was clear. 
Now, the structure of the modules is more complex: `mathics.builtin` has subpackages, and eventually sub-subpackages. 
Each module has also a docstring, which is associated with the introduction of a chapter or a guide.  Documentation has also now "Guide Sections", like in WR, which describe sets of symbols associated with certain tasks.

Maybe the problem here is we are trying to develop a documentation system that describes at the same time a hypermedia document (like WR) and a regular "linear" book, as well as a native "doctest" system. Let me explain what I understand looking at WR, and my reasons for proposing the new documentation structure organization. 

Structure of WR
=============

The hypermedial version fits more or less in the structure of WR: There are  "Guides" which references to "Builtin-reference pages". There is a "top level" (Reference part):
![imagen](https://github.com/Mathics3/mathics-core/assets/2862288/8a7b701b-688a-423d-a9be-123058344f9d)

Each entry  contains an index of "guides"
![imagen](https://github.com/Mathics3/mathics-core/assets/2862288/cee83e55-405c-4cfc-b2bd-8ccd47b2a747)

Each guide contains a short description and an index with links to other guides (titles in orange) and built-in reference for individual symbols.

The "built-in reference" has the form

![imagen](https://github.com/Mathics3/mathics-core/assets/2862288/7aafcc8b-60a2-47c3-818b-c551e9397c9d)

The same  "built-in reference" can appear in different guides.
![imagen](https://github.com/Mathics3/mathics-core/assets/2862288/fb14218d-3470-4d8b-b0a5-e46bad76dac4)

The appearance of "Guides" and "Built-in references" are similar, except for the kind of content itself and some color conventions.

If I would like to map the docstrings from modules and classes in `mathics.builtin` to this structure, it seems natural to associate modules docstrings with "Guides" and Builtin-symbols docstrings with "Built-in references". 

Chapters would then consist of related (like the ones associated to a pymathics module, or to a top-level module/package in mathics.builtin) Guides and Built-in references, which would be at the same documentation level (Sections).

Docpipeline
===========

Docpipeline makes two things: compare expressions obtained from Mathics evaluation with the expected results and produce the results to be shown in the documentation. For docpipeline, each test/result is identified with a key, composed of the part, chapter, and section titles, where the test belongs, and a numeric index. The proposed implementation of the documentation from WR is compatible with this because doctests always belong to one and just one chapter/ section.


PDF Book
=========

In the PDF book, we are forced to put elements sequentially, but we can play a little bit with the format. For example, we can decide to include just the Built-in reference sections and not the Guides, or add a TOC on each chapter, or use page numbers as references for each symbol in a Guide, or in the opposite way, keep just the guides and not the "Built-in reference" chapters like in "The Mathematica 5 Book".


So, why do I think that Built-in symbol references should not be mapped to subsections of a guide
==============================================================================

* The same Built-in symbol references entry can be associated with several guides.
* Built-in symbol references should be at the same level even if belong to a shallow module.
* In docpipeline, we associate Symbols with sections when referring to a set of tests.
* In WR, guides and references look very similar
* The scheme allows mapping each docstring into the documentation structure.
* gives a clear way to deal with nested packages.

**Reply by rocky** [reply]
> Maybe the problem here is we are trying to develop a documentation system that describes at the same time a hypermedia document (like WR) and a regular "linear" book, as well as a native "doctest" system. Let me explain what I understand looking at WR, and my reasons for proposing the new documentation structure organization.

Yes. There were very specific goals when I started undertaking this in preparation for a release, and the scope expanded to something totally different. Basically I wanted the documentation to be just like before, but with the ability  to be in a position to 

* support existing autoloading built-in functions. As with the import mechanism, the doc makes one pass gathering everything and has no support for incremental update after a builtin or set of builtins are autoloaded
* treat the handling of Mathics module the same way as a builtin. So a Mathics module starts off as a chapter like mathics.builtin modules do. (The module path of course is different, but that has to be). Right now, the Mathics3 Module system is organized differently (and less good) than the builtin system which allows for "guide sections". 

Side goals that I had hoped would fall out of this would be to be able to produce LaTeX documentation for a smaller section of the book so it can be debugged easier.  Also I didn't want to give up on the current behavior in Django where a dosctring can be updated and its effect is seen immediately.  The batch and monolithic behavior that we have largely inherited make working on things more tedious and slower and as a result, there are more doc mistakes and review due to the tediousness of building docs. 

Adding more URL tags like an image tag, doesn't feel important. I don't think we are in great need of adding static images. If I am wrong, let me know what images we need. Obviously I did not feel that way about adding URLs for cross referencing the document or to link to WMA or Wikipedia.  Overall, I think that has been a big win, at least for me.

Replicating the features of WR is also out of scope. It may be good to keep in mind where things might end up. But for now, there are far more important things to do. If I could just get those two items fixed (and it is not clear that right now they are), then I'd be happy. 

The coe module structure and books are l not hyperlinked . How one picks a tree structure from what is essentially a directed acyclic graph that WR has is arbitrary. That's just the way it is. 

> * The same Built-in symbol references entry can be associated with several guides.

This is easily addressed by adding  a class variable which lists all of the other guides. In terms of the code the symbol lives in one of guides and sections and subsections.

> * Built-in symbol references should be at the same level even if belong to a shallow module.

I don't agree at all, as far as the code goes. I am not sure I agree in a book either.  In Hyperlinked documentation, it is hard to tell whether something is at the same level.  In Django right now, I don't think anyone is bothered by the fact that from some menus may require another level to get to the symbol than some other symbols. One can get directly to a particular symbol by searching on the name of the symbol. And that's what I do when I want information about a symbol.

> * In docpipeline, we associate Symbols with sections when referring to a set of tests.

This doesn't bother me in the slightest. Just as with the Django doc another option can be added distinct from chapter or section to refer to a Symbol. 

> * In WR, guides and references look very similar

This might be a nice long-term thing to have but it wasn't on the short list of things to do before a release. Feel free to add this as a long-term goal.


> * The scheme allows mapping each docstring into the documentation structure.
> * gives a clear way to deal with nested packages.

I am not sure I understand exactly what these mean, but I don't think it matters. I'd like to defer all of this until after we get the next release out. None of this was in our roadmap.

**Reply by mmatera** [reply]
Ok, in that case, I will try to rewrite mathics.doc.gather to reproduce the documentation just as it was in the last release. Then we can continue the discussion and iterate over it.

**Reply by rocky** [reply]
If we could get the Mathics modules to work like builtins where they can have modules underneath that would be great too. So the Mathics3 Graphs Module can have Sections analogous to https://reference.wolfram.com/language/guide/GraphModifications.html

Also good would be cover the other bullet item where the documentation for a given module can be added lazily as will come up soon in autoloading.

Thanks.

**Reply by mmatera** [reply]
OK, #1022  now contains changes in #1025, but produces documentation with the format used in 6.0.

Also handles pymathics modules in a similar fashion as regular builtins: 

![imagen](https://github.com/Mathics3/mathics-core/assets/2862288/a918def0-547c-4a45-b60b-02da016edd72)

Each pymathics module like Graph is associated to a chapter. Each submodule produces a guide, containing the corresponding symbols as subsections.

Now the question is, what happens if `pymathics.graph.centralities` had subpackages? Should be subsubsections? we should "flatten" the packages, in a way that a subpackage produces a new guide?

**Reply by mmatera** [reply]
Notice also that `mathics.doc.gather.gather_reference_part` allows to process any list of modules to produce a "DocPart", so it would be possible to generate documentation from a Pymathics module, in a way that each module inside the package could produce a chapter.

**Comment by mmatera** [comment]
## Sphinx and readthedocs compatibility

Another possible improvement to the documentation system would be to use more standard syntax like CommonMarkdown or reStructuredText. Using a more standard notation would be helpful to replace the homebrew parser, based on Python regex, with another existing parser library. My first impulse was to consider Markdown, because of its simpler syntax, but @rocky has a point when he proposes using RsT, which has a more precise syntax and plays better with Sphinx, and seems to be the current standard. 

#1026 provides support for some basic notation like links, images, and code blocks in common Markdown. When I have another chunk of time, I will try to do the same to support RsT. 
Then, the long-term plan would be

* to support the most suitable syntax for using an existing documentation library
* Progressively replace the old syntax with the new one.
* Once the documentation can be handled by a standard parser, remove the homebrew, regex-based parser.
